File: ruby-todo/lib/engine.rb
module Engine
  def self.run(*args)
    run_with_history(*args).last
  end

  def self.run_with_history(model, messages)

  	new_model = model

    messages.map do |msg|
      new_model = msg.apply_to(new_model)
      new_model
    end
  end
end

File: swift-todo/Sources/Todo/Engine.swift
//
//  Engine.swift
//  Todo
//
//  Created by Paul on 2018/4/4.
//

import Foundation

struct Engine {
    static func run(on model: Model, applying messages: [Message]) -> Model {
        return runWithHistory(on: model, applying: messages).last!
    }

    static func runWithHistory(on model: Model, applying messages: [Message]) -> [Model] {
        var newModel = model
        return messages.map { message in
          newModel = message.apply(to: newModel)
          return newModel
        }
    }
}
